# Practice Session 01: Data preparation

Author: <font color="blue">Anna Quílez Rodríguez</font>

E-mail: <font color="blue"> anna.quilez01@estudiant.upf.edu</font>

AI used: <font color="blue">Chat GPT. I used artificial intelligence to help me understand the functions I wasn't familiar with. </font>

Date: <font color="blue">30/09/2026</font>

In [1]:
# LEAVE AS-IS

# Pandas library
import pandas as pd

# Show more rows in the output
pd.set_option("display.max_rows", 100)

# 1. Load dataset

In [2]:
# LEAVE AS-IS
input_dataset = pd.read_csv("data/airbnb/listings.csv.gz", compression="gzip")

# 2. Column types and null values <font color="blue">[10%]</font>

In [3]:
# LEAVE AS-IS

input_dataset.describe()

,id,scrape_id,neighborhood_overview,host_id,host_profile_id,host_since,hosts_time_as_user_years,hosts_time_as_user_months,hosts_time_as_host_years,hosts_time_as_host_months,...,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,instant_bookable,calculated_host_listings_count,calculated_host_listings_count_entire_homes,calculated_host_listings_count_private_rooms,calculated_host_listings_count_shared_rooms,reviews_per_month
count,1.529300e+04,1.529300e+04,0.0,1.529300e+04,1.520600e+04,0.0,15206.000000,15206.000000,15206.000000,15206.000000,...,11744.000000,11744.000000,11744.000000,11744.000000,0.0,15293.000000,15293.000000,15293.000000,15293.000000,11745.000000
mean,7.081310e+17,2.026062e+13,NaN,1.029976e+16,1.470058e+18,NaN,8.643628,5.134618,7.851309,4.990464,...,4.726335,4.722973,4.766025,4.465703,NaN,73.839992,65.734454,7.860394,0.124371,1.555447
std,6.550632e+17,0.000000e+00,NaN,1.316853e+17,2.707029e+16,NaN,4.120495,3.482615,4.282838,3.549561,...,0.463476,0.478455,0.362881,0.542243,NaN,146.257623,145.422963,30.752961,0.796254,2.318884
min,1.867400e+04,2.026062e+13,NaN,1.070400e+04,1.462507e+18,NaN,0.000000,0.000000,0.000000,0.000000,...,0.000000,1.000000,1.000000,1.000000,NaN,1.000000,0.000000,0.000000,0.000000,0.010000
25%,3.223180e+07,2.026062e+13,NaN,9.469830e+06,1.462659e+18,NaN,6.000000,2.000000,4.000000,1.000000,...,4.670000,4.670000,4.700000,4.330000,NaN,2.000000,1.000000,0.000000,0.000000,0.220000
50%,8.005535e+17,2.026062e+13,NaN,9.633738e+07,1.465487e+18,NaN,9.000000,5.000000,8.000000,5.000000,...,4.850000,4.860000,4.850000,4.570000,NaN,10.000000,5.000000,0.000000,0.000000,0.950000
75%,1.344730e+18,2.026062e+13,NaN,3.463675e+08,1.469532e+18,NaN,12.000000,8.000000,11.000000,8.000000,...,5.000000,5.000000,4.990000,4.770000,NaN,43.000000,31.000000,1.000000,0.000000,2.350000
max,1.714532e+18,2.026062e+13,NaN,1.713762e+18,1.714078e+18,NaN,17.000000,11.000000,15.000000,11.000000,...,5.000000,5.000000,5.000000,5.000000,NaN,588.000000,588.000000,205.000000,12.000000,100.580000


<font size="+1" color="blue">**Create function "describe_dataframe".** It must create and display a dataframe containing one row per column, and with the following fields: name of the column ("name"), type ("data_type"), number of non-null values ("count_not_null"), and number of distinct elements ("count_distinct").</font>

In [4]:
def describe_dataframe(df):

    # YOUR CODE HERE
    rows= []

    for column in df.columns:
      rows.append({"name": column,
            "data_type": df[column].dtype,
            "count_not_null": df[column].notnull().sum(),
            "count_distinct": df[column].nunique()})

    input_columns = pd.DataFrame(rows)

    return input_columns

input_columns = describe_dataframe(input_dataset)
display(input_columns)

,name,data_type,count_not_null,count_distinct
0,id,int64,15293,15293
1,listing_url,str,15293,15293
2,scrape_id,int64,15293,1
3,last_scraped,str,15293,4
4,source,str,15293,2
5,name,str,15293,14694
6,description,str,14644,11815
7,neighborhood_overview,float64,0,0
8,picture_url,str,15293,15145
9,host_id,int64,15293,4595


# 3. Remove empty columns <font color="blue">[10%]</font>

<font size="+1" color="blue">**Create function "drop_empty_columns"**. It should iterate through the columns in the original dataset, and return a new dataframe without the empty columns in the original one.</font>

In [5]:
def drop_empty_columns(df):

    # YOUR CODE HERE

    columns_to_keep = []

    for column in df.columns:
        if df[column].notnull().sum() != 0:
            columns_to_keep.append(column)

    return df[columns_to_keep].copy()

df1 = drop_empty_columns(input_dataset)

display(describe_dataframe(df1))

,name,data_type,count_not_null,count_distinct
0,id,int64,15293,15293
1,listing_url,str,15293,15293
2,scrape_id,int64,15293,1
3,last_scraped,str,15293,4
4,source,str,15293,2
5,name,str,15293,14694
6,description,str,14644,11815
7,picture_url,str,15293,15145
8,host_id,int64,15293,4595
9,host_url,str,15293,4595


# 4. Handle missing information about hosts <font color="blue">[30%]</font>

<font size="+1" color="blue">**Create function "drop_empty_hosts_v1"**. It should remove rows containing a null "host_profile_id" and indicate how many rows were removed (absolute and percentage).</font>

In [6]:
def drop_empty_hosts_v1(df):

    # YOUR CODE HERE
    total_rows= len(df)
    df_clean = df.dropna(subset=["host_profile_id"]).copy() #Drop rows with any column having NA/null data

    rows_removed = total_rows - len(df_clean)
    percentage_removed = rows_removed / total_rows * 100

    print("Rows removed:", rows_removed)
    print("Percentage removed:", percentage_removed, "%")

    return df_clean

df2 = drop_empty_hosts_v1(df1)

display(describe_dataframe(df2))

Rows removed: 87
Percentage removed: 0.5688877264107762 %


,name,data_type,count_not_null,count_distinct
0,id,int64,15206,15206
1,listing_url,str,15206,15206
2,scrape_id,int64,15206,1
3,last_scraped,str,15206,4
4,source,str,15206,2
5,name,str,15206,14637
6,description,str,14641,11812
7,picture_url,str,15206,15058
8,host_id,int64,15206,4553
9,host_url,str,15206,4553


<font size="+1" color="blue">**Write your strategy for handling null values in host-related columns**. Indicate in which cases you will drop the entire row, and in which cases you will replace by a default value.</font>

**Answer:** the strategy for handling nulls in host-related information will be as follows.

First, drop entire rows for nulls in these columns:
* Given that the total number of rows is 15,206 and host_profile_url has 15,203 non-null values, we can drop the rows with null values because they represent a very small minority of the total rows. Therefore, we are not losing much data. The prompt states that we can delete approximately 1% of the rows, so deleting these three rows is still below this limit. However, host_about and host_location have many null values, so we cannot delete these rows because we would lose too much information.

Second, replace with a default value the nulls in these columns:
* For host_location, we can replace the null values with the mode of the column because it represents the most common location. It is a reasonable default value for the missing locations.
            
Third, replace with a zero or empty string the nulls in these columns:
* For host_about, we can replace the null values with an empty string because this is a text field and there can be many different answers. Therefore, it is reasonable to leave it empty when the information is not available.

<font size="+1" color="blue">**Write a function that gives you the modal value for a column**. You will use to generate a default value.</font>

In [7]:
def determine_modal_value(df, column):
    
    # YOUR CODE HERE
    modal_value = df[column].mode()[0]
    
    return modal_value

<font size="+1" color="blue">**Write a function drop_empty_hosts_v2**. That does the same as its v1 plus applying your strategy for handling missing information about hosts.</font>

In [8]:
def drop_empty_hosts_v2(df):
    modal_location = determine_modal_value(df, "host_location")
    print(f"Modal location: {modal_location}")

    # YOUR CODE HERE
    total_rows = len(df)
    df_clean = df.dropna(subset=["host_profile_id", "host_profile_url"]).copy()

    rows_removed = total_rows - len(df_clean)
    percentage_removed = rows_removed / total_rows * 100

    print("Rows removed:", rows_removed)
    print("Percentage removed:", percentage_removed, "%")

    df_clean["host_location"] = df_clean["host_location"].fillna(modal_location)
    df_clean["host_about"] = df_clean["host_about"].fillna("")

    return df_clean

df2 = drop_empty_hosts_v2(df1)

display(describe_dataframe(df2))

Modal location: Barcelona, Spain
Rows removed: 90
Percentage removed: 0.5885045445628719 %


,name,data_type,count_not_null,count_distinct
0,id,int64,15203,15203
1,listing_url,str,15203,15203
2,scrape_id,int64,15203,1
3,last_scraped,str,15203,4
4,source,str,15203,2
5,name,str,15203,14634
6,description,str,14641,11812
7,picture_url,str,15203,15055
8,host_id,int64,15203,4553
9,host_url,str,15203,4553


In [9]:
# LEAVE AS-IS

# Check that all the host columns have the same number of values 
# as the id column
def check_host_columns(df):
    id_not_null = df["id"].notnull().sum()
    
    for col in df.columns:
        if col.startswith("host_"):
            null_values_in_column = df[col].notnull().sum()
            if null_values_in_column != id_not_null:
                print(f"Problem in column {col}, it has {null_values_in_column}")
                return False
            
    return True

check_host_columns(df2)

True

In [10]:
# LEAVE AS-IS

# Print table with columns containing null values
def null_value_counts(df):
    count_null_values = df.isnull().sum()
    columns_with_null_values = count_null_values[count_null_values > 0].reset_index()
    columns_with_null_values.columns = ["column", "null_count"]
    return columns_with_null_values
    
display(null_value_counts(df2))

,column,null_count
0,description,562
1,bathrooms,3475
2,bathrooms_text,13
3,bedrooms,2472
4,beds,2357
5,price,1926
6,price_quote_checkin_date,1620
7,price_quote_checkout_date,1620
8,price_quote_total_price,1926
9,price_quote_price_per_night,1926


# 5. Handle null values in number of bathrooms and bedrooms <font color="blue">[30%]</font>

<font size="+1" color="blue">**Include whatever code you need to decide your null-handling strategy for bathrooms and bedrooms**. Cross-tabulating the data is helpful, we have included one example of this.</font>

In [11]:
# Print distribution of the `bathrooms` column:
display(df2["bathrooms"].value_counts().sort_index())

# Print cross-distributions of the
# `bathrooms` and `bathrooms_text` columns.
# By default pd.crosstab drops null values, but we want to see them.

# AI Generated, edited (Claude Sonnet 5)
bathroom_crosstab = pd.crosstab(
    df2["bathrooms"].fillna("(null)"),
    df2["bathrooms_text"].fillna("(null)")
)
display(bathroom_crosstab)

# Continue with whatever you need to justify your decisions below
# YOUR CODE HERE
beds_bedrooms_crosstab = pd.crosstab(
    df2["beds"].fillna("(null)"),
    df2["bedrooms"].fillna("(null)")
)

display(beds_bedrooms_crosstab)

bathrooms
0.5       64
1.0     7053
1.5      540
2.0     3260
2.5      141
3.0      291
3.5       27
4.0      164
4.5        9
5.0       57
5.5        5
6.0       45
7.0        6
8.0       40
9.0        4
10.0      10
11.0       3
12.0       6
13.0       1
14.0       1
16.0       1
Name: count, dtype: int64

bathrooms_text,(null),0 baths,0 shared baths,1 bath,1 private bath,1 shared bath,1.5 baths,1.5 shared baths,10 baths,11 baths,...,5.5 baths,6 baths,7 baths,7 shared baths,8 baths,8 shared baths,9 baths,Half-bath,Private half-bath,Shared half-bath
bathrooms,,,,,,,,,,,,,,,,,,,,,
0.5,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,21,7,36
1.0,0,1,0,5310,783,959,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1.5,1,0,0,0,0,0,410,129,0,0,...,0,0,0,0,0,0,0,0,0,0
2.0,1,0,0,1,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2.5,1,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3.0,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3.5,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4.0,1,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4.5,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


bedrooms,0.0,1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0,9.0,...,14.0,15.0,16.0,18.0,19.0,20.0,24.0,26.0,29.0,(null)
beds,,,,,,,,,,,,,,,,,,,,,
1.0,0,2802,72,19,9,9,1,4,9,3,...,0,0,0,0,1,0,0,0,1,1220
2.0,0,1046,1611,16,2,0,0,0,0,1,...,0,0,0,0,0,0,0,0,0,153
3.0,0,224,1438,781,3,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,32
4.0,0,62,527,766,247,1,0,1,0,0,...,0,0,0,0,0,0,0,0,0,14
5.0,0,9,111,431,149,39,0,0,0,0,...,0,0,0,0,0,0,0,0,0,7
6.0,0,12,8,228,151,24,12,0,0,0,...,0,0,0,0,0,0,0,0,0,5
7.0,0,3,4,57,98,33,8,1,0,0,...,0,0,0,0,0,0,0,0,0,3
8.0,0,11,2,5,36,22,10,1,7,2,...,0,0,0,0,0,0,0,0,0,0
9.0,0,0,1,3,11,13,22,0,1,6,...,0,0,0,0,0,0,0,0,0,0


<font size="+1" color="blue">**Write your strategy for handling null values in bathrooms and bedrooms**. Justify briefly your choices.</font>

**Answer:** the strategy for handling missing values in `bathrooms`, `bathrooms_text`, `bedrooms` and `beds` will be as follows.

For columns `bathrooms` and `bathroom_text`:
* If bathrooms is null but bathrooms_text contains the number of bathrooms, we will use bathrooms_text to fill bathrooms. If bathrooms_text is null but bathrooms has a value, we will generate the corresponding text from bathrooms.

For columns `bedrooms` and `beds`:
* if bedrooms is null but beds is known, we will calculate bedrooms according to the given rule: 1 bed means 0 bedrooms, and 2 or more beds means beds - 1 bedrooms. If beds is missing we use the number of bedrooms. If both are missing, we remove the row because we cannot know the correct values.

<font size="+1" color="blue">**Implement your strategy for handling null values in bathrooms and bedrooms**. </font>

In [12]:
def handle_null_values_bathrooms_bedrooms(df):
    df_clean = df.copy()

    # null bathrooms
    for i in df_clean.index:
        if pd.isnull(df_clean.loc[i, "bathrooms"]):
            text = df_clean.loc[i, "bathrooms_text"]

            if text in ["Half-bath", "Shared half-bath", "Private half-bath"]:
                df_clean.loc[i, "bathrooms"] = 0.5
            elif not pd.isnull(text):
                df_clean.loc[i, "bathrooms"] = float(text.split()[0])

    # null bathrooms text
    for i in df_clean.index:
        if pd.isnull(df_clean.loc[i, "bathrooms_text"]):
            df_clean.loc[i, "bathrooms_text"] = str(df_clean.loc[i, "bathrooms"]) + " baths"

    # null bedrooms
    for i in df_clean.index:
        if pd.isnull(df_clean.loc[i, "bedrooms"]):
            if df_clean.loc[i, "beds"] == 1:
                df_clean.loc[i, "bedrooms"] = 0
            elif df_clean.loc[i, "beds"] >= 2:
                df_clean.loc[i, "bedrooms"] = df_clean.loc[i, "beds"] - 1

    # null beds
    for i in df_clean.index:
        if pd.isnull(df_clean.loc[i, "beds"]):
            df_clean.loc[i, "beds"] = df_clean.loc[i, "bedrooms"]

    # Remove rows where we still do not have enough information
    df_clean = df_clean.dropna(subset=["bathrooms", "bedrooms", "beds"])

    return df_clean


df3 = handle_null_values_bathrooms_bedrooms(df2)

# This should show no NULL values in bathrooms or bedrooms
display(null_value_counts(df3))

,column,null_count
0,description,499
1,price,1032
2,price_quote_checkin_date,741
3,price_quote_checkout_date,741
4,price_quote_total_price,1032
5,price_quote_price_per_night,1032
6,price_quote_raw,741
7,minimum_nights,1
8,maximum_nights,1
9,minimum_minimum_nights,1


# 6. Handle null values in other columns <font color="blue">[20%]</font>

<font size="+1" color="blue">**Write your strategy for the remaining columns with null values**. Indicate for each column if you will (a) drop the rows containig null values, (b) replace by a modal value, (c) replace by a zero or empty string. Justify briefly each choice.</font>

**Answer:** the strategy for handling missing values will be as follows.

Columns related to the description of the property:

* Column `description`: replace with an empty string, as an empty description is valid.

Columns related to minimum and maximum number of nights (`*_nights`):

* These columns are minimum_nights, maximum_nights, minimum_minimum_nights, maximum_minimum_nights, minimum_maximum_nights and maximum_maximum_nights. We will drop the rows containing null values. As we can see, there is only 1 null value in each of these columns, therefore, removing these rows will have a minimal impact and we will not lose much data. Moreover, it does not make sense to replace these values with the mode because we would be inventing a value, and putting 0 would be invalid because we cannot stay a minimum of 0 nights.

Columns related to availability, price, and license:

* These columns are price, price_quote_checkin_date, price_quote_checkout_date, price_quote_total_price, price_quote_price_per_night, price_quote_raw, has_availability, estimated_revenue_l365d and license. Drop the rows. We can not use a listing if we don't know if it is available, its price, or if it has a license.


Columns related to reviews:

* For review scores these, columns are review_scores_rating, review_scores_accuracy, review_scores_cleanliness, review_scores_checkin, review_scores_communication, review_scores_location and review_scores_value. We would replace them with the average value of each column. Whereas for the missing values in reviews_per_month, we put a 0 because a listing with no reviews has zero reviews per month. Furthermore, first_review as well as last_review will remain null, given that they are dates and a null value means that the listing has never been reviewed.

<font size="+1" color="blue">**Implement your strategy for handling the remaining null values**. </font>

In [13]:
def handle_remaining_null_values(df):
    df_clean = df.copy()
    
    # YOUR CODE HERE
    # Description 
    df_clean["description"] = df_clean["description"].fillna("")
    
    # Minimum and maximum 
    df_clean = df_clean.dropna(subset=["minimum_nights","maximum_nights","minimum_minimum_nights", "maximum_minimum_nights","minimum_maximum_nights","maximum_maximum_nights"])

    # Availability, price and license 
    df_clean = df_clean.dropna(subset=["has_availability", "price","price_quote_checkin_date","price_quote_checkout_date","price_quote_total_price","price_quote_price_per_night","price_quote_raw","estimated_revenue_l365d","license"])
    # Review scores 
    review_columns = ["review_scores_rating","review_scores_accuracy","review_scores_cleanliness","review_scores_checkin", "review_scores_communication","review_scores_location","review_scores_value"]

    for column in review_columns:
        average_value = df_clean[column].mean()
        df_clean[column] = df_clean[column].fillna(average_value)
        
    # Reviews 
    df_clean["reviews_per_month"] = df_clean["reviews_per_month"].fillna(0)
    df_clean["first_review"] = df_clean["first_review"].fillna("")
    df_clean["last_review"] = df_clean["last_review"].fillna("")
    
    return df_clean
    
df4 = handle_remaining_null_values(df3)

# This should show no column with a null value
display(null_value_counts(df4))

,column,null_count


In [14]:
# LEAVE AS-IS

# Save the data, we will use it in the next assignment
df4.to_csv("listings_clean.csv.gz", index=False, compression="gzip")

<font size="+2" color="#003300">I hereby declare that I completed this practice myself, that I have not copied it from others, or from external sources without citing them, and that I have followed the generative AI instructions stated above.</font>